### DISCUSSION TITLE
explanation of example 12

### DISCUSSION THREAD

**Discussion by joeb8799** [NatLabRockies/ROSCO#140] [General]
Hello, 
Can someone explain for me example 12, i need to understand the meaning of stability margin and how it is calculated.

**Comment by nikhar-abbas** [comment]
Hi,

Example 12 performs a stability-constrained tuning similar to that presented in the following paper:
https://onlinelibrary.wiley.com/doi/full/10.1002/we.2408

We have another paper that is currently under review that provides more details on the specifics of how we tune ROSCO accordingly, but it cannot be shared publicly yet. 

Here is a high-level description of the inputs:
The stability constrained method is mostly defined by the inputs to the `opt_options` [here](https://github.com/NREL/ROSCO/blob/77278f5d1cee95a830b22d79797b351481e87972/Examples/example_12.py#L67). A value of `omega_pc` schedule is calculated for each windspeed `linmodel_tuning['U_pc']` such that the stability margin `linmodel_tuning['stability_margin']` is not violated, and that `omega_pc` is not larger than `linmodel_tuning['omega_pc']['max']`.

Converting this do a discussion b/c it doesn't seem like this is an issue.

-Nikhar

**Comment by bkwang18** [comment]
hello, 

I met an error when I tried to run example_12, it said "Number of blades is  0  not 3. ", but I have checked the input file, it has defined with 3 blades, then the error happened, I don't know whether the error is related to the above message or not.
Is there anyone who can give me some hint about this example? Thanks. 


PS D:\ROSCO> & D:/Anaconda3/envs/rosco-env/python.exe d:/ROSCO/Examples/example_12.py
Using weis.aeroelasticse in ROSCO_toolbox...
-----------------------------------------------------------------------------
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox    
-----------------------------------------------------------------------------
Loading FAST model: IEA-15-240-RWT-UMaineSemi.fst 
Loading rotor performace data from text file: d:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
Loading rotor performace data from text file: d:\ROSCO\Examples\../Test_Cases/IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
-----------------------------------------------------------------------------
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox    
-----------------------------------------------------------------------------
Loading FAST model: IEA-15-240-RWT-UMaineSemi.fst
Loading rotor performace data from text file: d:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_00
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_02
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_04
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_06
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_05
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_01
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_07
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_03
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_08
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_10
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_12
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_14
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_09
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_11
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_13
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_15
    raise self._exc_info[1].with_traceback(self._exc_info[2])
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\drivers\scipy_optimizer.py", line 688, in _gradfunc
    grad = self._compute_totals(of=self._obj_and_nlcons, wrt=self._dvlist,
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\utils\hooks.py", line 131, in execute_hooks
    _run_hooks(post_hooks, inst)
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\utils\hooks.py", line 107, in _run_hooks
    hook(inst, **kwargs)
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\visualization\scaling_viewer\scaling_report.py", line 582, in _run_scaling_report
    prob.driver.scaling_report(outfile=scaling_filepath, show_browser=False)
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\core\driver.py", line 1139, in scaling_report
    return view_driver_scaling(self, outfile=outfile, show_browser=show_browser, jac=jac,
  File "D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\visualization\scaling_viewer\scaling_report.py", line 456, in view_driver_scaling
    d3_src = f.read()
UnicodeDecodeError: 'gbk' codec can't decode byte 0x92 in position 101887: illegal multibyte sequence

**Reply by dzalkind** [reply]
The warning `fx_mbc3 WARNING: Number of blades is 0 not 3. MBC transformation was not performed.` is because there are no blade states in the linear model.  I don't think this is affecting the error at the end.  I suspect this might be an openmdao version issue.  Do you know what version you are using?  You can find out by entering `conda list | grep openmdao`

**Reply by bkwang18** [reply]
hi,

my current openmdao version is below, is this OK? Thanks.

openmdao                  3.19.0             pyhd8ed1ab_0    conda-forge

**Reply by dzalkind** [reply]
Can you try version 3.18.0?  `conda install openmdao==3.18.0`. If that works, we'll update the dependency.  Thanks!

**Comment by bkwang18** [comment]
Sorry, it didn't work, and you can see the bottom, I have switched the
version of openmdao to 3.18.0 already.

D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\recorders\sqlite_recorder.py:227:
UserWarning:The existing case recorder file,
d:\ROSCO\Examples\examples_out\12_robust_scheduling.12.opt.sql, is being
overwritten.
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135:
OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is
currently scaled by the norm of the vector variable. This is not ideal for
larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To
preserve the older way of doing this calculation, set step_calc to
'rel_legacy'.
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135:
OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is
currently scaled by the norm of the vector variable. This is not ideal for
larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To
preserve the older way of doing this calculation, set step_calc to
'rel_legacy'.
Traceback (most recent call last):
  File "d:/ROSCO/Examples/example_12.py", line 128, in <module>
    run_example()
  File "d:/ROSCO/Examples/example_12.py", line 86, in run_example
    sd.execute()
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\rosco-2.5.0-py3.8-win-amd64.egg\ROSCO_toolbox\linear\robust_scheduling.py",
line 230, in execute
    self.om_opt.run_driver()
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\core\problem.py",
line 767, in run_driver
    return self.driver.run()
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\drivers\scipy_optimizer.py",
line 532, in run
    self._reraise()
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\drivers\scipy_optimizer.py",
line 778, in _reraise
    raise self._exc_info[1].with_traceback(self._exc_info[2])
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\drivers\scipy_optimizer.py",
line 700, in _gradfunc
    grad = self._compute_totals(of=self._obj_and_nlcons, wrt=self._dvlist,
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\utils\hooks.py",
line 127, in execute_hooks
    _run_hooks(post_hooks, inst)
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\utils\hooks.py",
line 101, in _run_hooks
    hook(inst, **kwargs)
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\utils\reports_system.py",
line 314, in run_scaling_report_inner
    prob.driver.scaling_report(outfile=scaling_filepath, show_browser=False)
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\core\driver.py",
line 1131, in scaling_report
    return view_driver_scaling(self, outfile=outfile,
show_browser=show_browser, jac=jac,
  File
"D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\visualization\scaling_viewer\scaling_report.py",
line 457, in view_driver_scaling
    d3_src = f.read()
UnicodeDecodeError: 'gbk' codec can't decode byte 0x92 in position 101887:
illegal multibyte sequence
PS D:\ROSCO> conda list openmdao
# packages in environment at D:\Anaconda3\envs\rosco-env:
#
# Name                    Version                   Build  Channel
openmdao                  3.18.0             pyhd8ed1ab_0    conda-forge

On Fri, Jul 22, 2022 at 7:28 PM dzalkind ***@***.***> wrote:

> Can you try version 3.18.0? conda install openmdao==3.18.0. If that
> works, we'll update the dependency. Thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NREL/ROSCO/discussions/140#discussioncomment-3208054>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADFK7UEAO66ZCIO4BZMDPFTVVLK5LANCNFSM54HUIREQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Bingkun Wang

**Reply by bkwang18** [reply]
I change the openmdao version back to v3.12.0, then I think it should be working with message below:

Using weis.aeroelasticse in ROSCO_toolbox...
-----------------------------------------------------------------------------
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox
-----------------------------------------------------------------------------
Loading FAST model: IEA-15-240-RWT-UMaineSemi.fst 
Loading rotor performace data from text file: d:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
Loading rotor performace data from text file: d:\ROSCO\Examples\../Test_Cases/IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
-----------------------------------------------------------------------------
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox
-----------------------------------------------------------------------------
Loading FAST model: IEA-15-240-RWT-UMaineSemi.fst
Loading rotor performace data from text file: d:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\Cp_Ct_Cq.IEA15MW.txt
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Using weis.aeroelasticse in ROSCO_toolbox...
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_00
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_02
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_04
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_06
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_01
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_03
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_05
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_07
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_08
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_10
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_12
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_14
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_11
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_09
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_13
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_15
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_16
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_18
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_17
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
Loading linearizations from: D:\ROSCO\Test_Cases\IEA-15-240-RWT-UMaineSemi\linearizations\lin_19
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
 fx_mbc3 WARNING: Number of blades is  0  not 3. MBC transformation was not performed.
loaded in parallel in 14.438570737838745 seconds
Finding ROSCO tuning parameters for u = 12, sm = 0.1
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\recorders\sqlite_recorder.py:223: UserWarning:The existing case recorder file, d:\ROSCO\Examples\examples_out\12_robust_scheduling.12.opt.sql, is being overwritten.
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.08028662174921605
            Iterations: 3
            Function evaluations: 5
            Gradient evaluations: 3
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 14, sm = 0.1
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.1210334602636735
            Iterations: 2
            Function evaluations: 3
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 15, sm = 0.1
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.22217151329015056
            Iterations: 5
            Function evaluations: 6
            Gradient evaluations: 5
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 18, sm = 0.1
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete
-----------------------------------
Finding ROSCO tuning parameters for u = 20, sm = 0.1
D:\Anaconda3\envs\rosco-env\lib\site-packages\openmdao\approximation_schemes\finite_difference.py:135: OMDeprecationWarning:When using 'rel' as the step_calc, the fd stepsize is currently scaled by the norm of the vector variable. This is not ideal for larger vectors, and this behavior is being changed in OpenMDAO 3.12.0. To preserve the older way of doing this calculation, set step_calc to 'rel_legacy'.
Optimization terminated successfully    (Exit mode 0)
            Current function value: -0.3
            Iterations: 2
            Function evaluations: 2
            Gradient evaluations: 2
Optimization Complete

**Reply by dzalkind** [reply]
Thanks!  We'll revert to 3.12.0.  I learned there is an issue with the openmdao logging in newer versions. 

If you want to use a newer version, the quick workaround is to disable reports by setting an environment variable: [http://openmdao.org/twodocs/versions/latest/features/reports/reports_system.html#controlling-reporting-using-an-environment-variable](https://gcc02.safelinks.protection.outlook.com/?url=http%3A%2F%2Fopenmdao.org%2Ftwodocs%2Fversions%2Flatest%2Ffeatures%2Freports%2Freports_system.html%23controlling-reporting-using-an-environment-variable&data=05%7C01%7CDaniel.Zalkind%40nrel.gov%7C291261f6af6a4abeafe208da6c1667d9%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C637941140744464187%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=8DhsMl600CtuMX70Q9mGuaTGIH4nZNGHGIJ%2FnfeHk60%3D&reserved=0)
